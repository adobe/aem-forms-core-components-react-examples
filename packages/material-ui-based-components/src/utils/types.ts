import { FieldJson, FieldsetJson, State } from '@aemforms/af-core';
import { Handlers, WithViewState } from '@aemforms/af-react-renderer';

export type FieldViewState = WithViewState<FieldJson>;

export type PROPS = State<FieldJson & Handlers & {
  isError?: boolean,
  layout?: {
    [key: string]: any
  }
  text?:string,
  richText?:boolean
}>;

export type PROPS_PANEL = State<FieldsetJson> & { handlers: Handlers }
