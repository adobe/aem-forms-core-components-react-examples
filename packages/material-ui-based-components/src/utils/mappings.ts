import Form from '../components/Form';
import Panel from '../components/Panel';
import TextField from '../components/TextField';
import Date from '../components/Date';
import PlainText from '../components/PlainText';
import Title from '../components/Title';
import Image from '../components/Image';
import Rating from '../components/Rating';
import Slider from '../components/Slider';
import TabsVerticalComp from '../components/TabsVerticalComp';
import TabsHorizontalComp from '../components/TabsHorizontalComp';

export default {
  form: Form,
  panel: Panel,
  'text-input': TextField,
  'date-input': Date,
  'text-input': TextField,
  'plain-text': PlainText,
  'custom:title': Title,
  image: Image,
  'text-input': TextField,
  'custom:rating':Rating,
  'custom:slider':Slider
  'text-input': TextField,
  'custom:vertical-tab': TabsVerticalComp,
  'custom:horizontal-tab': TabsHorizontalComp
};