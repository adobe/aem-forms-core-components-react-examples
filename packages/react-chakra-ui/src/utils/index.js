import React from 'react';
import { FormContext } from '@aemforms/af-react-renderer';
import { createFormInstance } from '@aemforms/af-core';
import { ChakraProvider } from '@chakra-ui/react'
import { IntlProvider } from 'react-intl';
import { render } from '@testing-library/react';

export const DEFAULT_ERROR_MESSAGE = 'There is an error in the field';

export const createForm = (field) => {
  const formJson = {
    items: [field]
  };
  return createFormInstance(formJson);
};

export const Provider = (form, mappings, locale, dictionaries) => props => {
  const c = {
    form,
    mappings,
    modelId: '$form',
    refMap: {}
  };
  const { children } = props;
  return (
    <ChakraProvider>
      <IntlProvider locale={locale} messages={dictionaries}>
        <FormContext.Provider value={c}>{children}</FormContext.Provider>
      </IntlProvider>
    </ChakraProvider>
  );
};

export const renderComponent = function (Component) {
  const test = (field, operation) => {
    const form = createForm(field);
    if (operation) {
      operation(form, form.items[0]);
    }
    const e = form.items[0].getState();
    let component = <Component {...e} />;
    const wrapper = Provider(form);
    const renderResponse = render(component, { wrapper });
    return {
      renderResponse,
      form,
      element: form?.items[0],
      component
    };
  };
  return test;
};
