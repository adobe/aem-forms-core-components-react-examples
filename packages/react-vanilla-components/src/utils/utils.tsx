export const formatBytes = (bytes: number, decimals = 0) => {
    if (!+bytes) {
      return '0 Bytes';
    }
    const k = 1024;
    const dm = decimals < 0 ? 0 : decimals;
    const sizes = ['bytes', 'kb', 'mb', 'gb', 'tb', 'pb'];
    const i = Math.floor(Math.log(bytes) / Math.log(k));

    return `${parseFloat((bytes / Math.pow(k, i)).toFixed(dm))} ${sizes[i]}`;
  };

  export const syncAriaDescribedBy = (id: string, tooltip?: string, description?: string, error?: string) => {
    const descriptions = [];

    if (description) {
        descriptions.push(`${id}__longdescription`);
    }
    if (tooltip) {
        descriptions.push(`${id}__shortdescription`);
    }
    if (error) {
        descriptions.push(`${id}__errormessage`);
    }

    return descriptions.join(' ');
};