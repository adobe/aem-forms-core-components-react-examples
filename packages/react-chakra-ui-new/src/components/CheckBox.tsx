/*************************************************************************
* ADOBE CONFIDENTIAL
* ___________________
*
* Copyright 2023 Adobe
* All Rights Reserved.
*
* NOTICE: All information contained herein is, and remains
* the property of Adobe and its suppliers, if any. The intellectual
* and technical concepts contained herein are proprietary to Adobe
* and its suppliers and are protected by all applicable intellectual
* property laws, including trade secret and copyright laws.
* Dissemination of this information or reproduction of this material
* is strictly forbidden unless prior written permission is obtained
* from Adobe.

* Adobe permits you to use and modify this file solely in accordance with
* the terms of the Adobe license agreement accompanying it.
*************************************************************************/

import React, { useCallback } from 'react';
import { withRuleEngine } from '../shared/withRuleEngine';
import { PROPS } from '../utils/type';
import { FormControl, Checkbox, FormHelperText } from '@chakra-ui/react';

const CheckBoxComponent = (props: PROPS) => {
    const {
        label, required, enabled, value, errorMessage,
        description, isError, name
    } = props;
    const selectedValue = props.enum?.[0];
    const unselectedValue = (props.enum?.length || 0) < 2 ? null : props.enum?.[1];

    const changeHandler = useCallback((event: any) => {
        const val = (event.target.checked) ? selectedValue : unselectedValue;
        props.dispatchChange(val);
    }, [props.dispatchChange]);

    return (
        <FormControl
            variant={props.layout?.variant}
            isRequired={required}
            isDisabled={!enabled}
        >
            <Checkbox
                {...props?.properties?.['afs:layout']}
                isDisabled={!enabled}
                isRequired={required}
                value={label}
                onChange={changeHandler}
                name={name}
                isChecked={value === props.enum?.[0] || value === true}
            >
                {label?.visible ? label.value : ''}
            </Checkbox>
            <FormHelperText>
                {isError ? errorMessage : description}
            </FormHelperText>
        </FormControl>
    );
};

export default withRuleEngine(CheckBoxComponent);