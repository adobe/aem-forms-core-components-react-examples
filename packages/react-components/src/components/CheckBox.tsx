import React from 'react';
import {State, FieldJson} from '@aemforms/af-core';
import { IHandler } from '../utils/type';
import withRuleEngine from '../utils/HOC';

const CheckBox = (props: State<FieldJson> & IHandler) => {
   const { id, label, enum: enums,  value, valid, errorMessage, description, name, readOnly, enabled } = props;  
   const selectedValue = enums?.[0];
   const unSelectedValue = (enums?.length || 0) < 2 ? null : enums?.[1];
   const validateState = valid === false ? 'invalid' : 'valid';
   const error = validateState === 'invalid';
   const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const val = e.target.checked ? selectedValue : unSelectedValue;
      props.dispatchChange(val);
   };
   return (
     <div className="cmp-adaptiveform-checkbox">
       <input type='checkbox' className="cmp-adaptiveform-checkbox__widget" onChange={handleChange} value={value} name={name} readOnly={readOnly} disabled={!enabled} />
       {label?.visible && <label className="cmp-adaptiveform-checkbox__label">{label?.value}</label>}
       {error && <div id={`${id}-errorMessage`} className="cmp-adaptiveform-checkbox__errormessage">{errorMessage}</div>}
       {description && !error && <div id={`${id}-longDescription`} className="cmp-adaptiveform-checkbox__longdescription">{description}</div>}
     </div>
   );   
};

export default withRuleEngine(CheckBox);