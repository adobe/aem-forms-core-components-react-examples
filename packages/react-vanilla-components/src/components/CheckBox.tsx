// *******************************************************************************
//  * Copyright 2023 Adobe
//  *
//  * Licensed under the Apache License, Version 2.0 (the “License”);
//  * you may not use this file except in compliance with the License.
//  * You may obtain a copy of the License at
//  *
//  *     http://www.apache.org/licenses/LICENSE-2.0
//  *
//  * Unless required by applicable law or agreed to in writing, software
//  * distributed under the License is distributed on an “AS IS” BASIS,
//  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  * See the License for the specific language governing permissions and
//  * limitations under the License.
//  ******************************************************************************

import React, { useState } from 'react';
import { withRuleEngine } from '../utils/withRuleEngine';
import { PROPS } from '../utils/type';

const CheckBox = (props: PROPS) => {
  const { id, label, enum: enums, value, isError, errorMessage, description, name, readOnly, enabled, visible } = props;
  const selectedValue = enums?.[0];
  const [shortDescription, setShortDescription] = useState(true);
  const [longDescription, setLongtDescription] = useState(false);
  const unSelectedValue = (enums?.length || 0) < 2 ? null : enums?.[1];

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const val = e.target.checked ? selectedValue : unSelectedValue;
    props.dispatchChange(val);
  };

  const handleClick = (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
    event.preventDefault();
    setShortDescription(!shortDescription);
    setLongtDescription(!longDescription);
  };

  return (
    <div className="cmp-adaptiveform-checkbox" data-cmp-is="adaptiveFormCheckBox" data-cmp-visible={visible} data-cmp-enabled={enabled}>
      {label?.visible && <label className="cmp-adaptiveform-checkbox__label" htmlFor={`${id}-widget`}>{label?.value}</label>}
      {description && <button aria-label='Toggle Button' className="cmp-adaptiveform-checkbox__questionmark" onClick={handleClick}></button>}
      <input id={`${id}-widget`} type='checkbox' className={selectedValue === value ? 'cmp-adaptiveform-checkbox__widget cmp-adaptiveform-checkbox__widget--filled' : 'cmp-adaptiveform-checkbox__widget cmp-adaptiveform-checkbox__widget--empty'} onChange={handleChange} value={value} name={name} readOnly={readOnly} disabled={!enabled} aria-checked={selectedValue === value ? 'true' : 'false'} />
      {shortDescription && props?.tooltip && <div title='Help Text' className='cmp-adaptiveform-checkbox__shortdescription' data-cmp-visible={shortDescription}>{props?.tooltip}</div>}
      <div aria-live="polite">
        {longDescription && description && !errorMessage ? <div title='Help Text' data-cmp-visible={longDescription} className="cmp-adaptiveform-checkbox__longdescription">{description}</div> : null}
      </div>
      {isError ? <div id={`${id}-errorMessage`} className="cmp-adaptiveform-checkbox__errormessage">{errorMessage}</div> : null}
    </div>
  );
};

export default withRuleEngine(CheckBox);