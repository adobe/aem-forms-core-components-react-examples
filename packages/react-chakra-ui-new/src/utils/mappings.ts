// *******************************************************************************
//  * Copyright 2023 Adobe
//  *
//  * Licensed under the Apache License, Version 2.0 (the “License”);
//  * you may not use this file except in compliance with the License.
//  * You may obtain a copy of the License at
//  *
//  *     http://www.apache.org/licenses/LICENSE-2.0
//  *
//  * Unless required by applicable law or agreed to in writing, software
//  * distributed under the License is distributed on an “AS IS” BASIS,
//  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  * See the License for the specific language governing permissions and
//  * limitations under the License.
//  ******************************************************************************

import Button from '../components/Button';
import CheckBox from '../components/CheckBox';
import CheckBoxGroup from '../components/CheckBoxGroup';
import DropDown from '../components/DropDown';
import Number from '../components/Number';
import RadioGroup from '../components/RadioButton';
import TextField from '../components/TextField';
import Panel from '../components/Panel';
import Wizard from '../components/Wizard';
import Accordion from '../components/Accordion';
import HorizontalTab from '../components/HorizontalTabs';
import VerticalTab from '../components/VerticalTabs';
import Form from '../components/Form';
import Image from '../components/Image';
import PlainText from '../components/PlainText';

const mappings = {
  'text-input': TextField,
  'password-input': TextField,
  'multiline-input': TextField,
  panel: Panel,
  'checkbox-group': CheckBoxGroup,
  'radio-group': RadioGroup,
  'drop-down': DropDown,
  'number-input': Number,
  button: Button,
  checkbox: CheckBox,
  image: Image,
  'core/fd/components/form/wizard/v1/wizard': Wizard,
  'core/fd/components/form/accordion/v1/accordion': Accordion,
  'core/fd/components/form/tabsontop/v1/tabsontop': HorizontalTab,
  'core/fd/components/form/verticaltabs/v1/verticaltabs': VerticalTab,
  'plain-text': PlainText,
  'core/fd/components/form/panelcontainer/v1/panelcontainer': Panel,
  form: Form
};

export default mappings;