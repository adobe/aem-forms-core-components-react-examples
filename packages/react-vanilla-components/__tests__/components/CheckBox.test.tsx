/*
 * Copyright 2023 Adobe, Inc.
 *
 * Your access and use of this software is governed by the Adobe Customer Feedback Program Terms and Conditions or other Beta License Agreement signed by your employer and Adobe, Inc.. This software is NOT open source and may not be used without one of the foregoing licenses. Even with a foregoing license, your access and use of this file is limited to the earlier of (a) 180 days, (b) general availability of the product(s) which utilize this software (i.e. AEM Forms), (c) January 1, 2023, (d) Adobe providing notice to you that you may no longer use the software or that your beta trial has otherwise ended.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL ADOBE NOR ITS THIRD PARTY PROVIDERS AND PARTNERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import CheckBox from '../../src/components/CheckBox';
import { renderComponent } from '../utils';
import userEvent from "@testing-library/user-event";

const field = {
  name: 'checkbox',
  label: {
    value: 'CheckBox Button',
    visible: true
  },
  fieldType: 'checkbox',
  visible: true,
  type: 'boolean',
};

const helper = renderComponent(CheckBox);

describe('Checkbox', () => {

  test('Snapshot', async () => {
    const { renderResponse } = await helper(field);
    expect(renderResponse.container).toMatchSnapshot();
  });

  test('selection made by the user sets the value', async () => {
    const f = {
      ...field,
    };
    const { renderResponse, element } = await helper(f);
    const input = await renderResponse.container.getElementsByClassName("cmp-adaptiveform-checkbox__widget");
    userEvent.click(input[0]);
    expect(element?.value).toEqual(true);
  });

  test('clicking on checkbox twice resets the value', async () => {
    const f = {
      ...field,
      enum: [false, true],
      value: [true, false]
    };
    const { renderResponse, element } = await helper(f);
    const input = await renderResponse.container.getElementsByClassName("cmp-adaptiveform-checkbox__widget");
    userEvent.click(input[0]);
    userEvent.click(input[0]);
    expect(element?.getState().value).toEqual(element.value);
  });

  test('a checkbox with no off value should get its value undefined when not selected', async () => {
    const f = {
      ...field,
      enum: [true],
    };
    const { renderResponse, element } = await helper(f);
    const input = await renderResponse.container.getElementsByClassName("cmp-adaptiveform-checkbox__widget");
    userEvent.click(input[0]);
    let state = element?.getState();
    expect(state.value).toBe(true);
    userEvent.click(input[0]);
    state = element?.getState();
    expect(state.value).toBe(null);
  });

  test('a checkbox with no off value should be valid when not required', async () => {
    const f = {
      ...field,
      enum: [true],
    };
    const { renderResponse, form, element } = await helper(f);
    const input = await renderResponse.container.getElementsByClassName("cmp-adaptiveform-checkbox__widget");
    userEvent.click(input[0]);
    let data = form?.exportData();
    expect(data.checkbox).toEqual(true);
    userEvent.click(input[0]);
    expect(element?.valid).not.toBe(false);
  });

  test('a checkbox should render if there are no options', async () => {
    const { renderResponse, element } = await helper({ ...field });
    const input = await renderResponse.container.getElementsByClassName("cmp-adaptiveform-checkbox__widget");
    userEvent.click(input[0]);
    userEvent.click(input[0]);
    expect(element?.getState().valid).toBe(true);
  });

  test('a checkbox should be selected if value is on', async () => {
    const f = {
      ...field,
      value: true,
    };
    const { renderResponse, element } = await helper(f);
    const input = await renderResponse.container.getElementsByClassName("cmp-adaptiveform-checkbox__widget");
    expect(input).not.toBeNull();
    let state = element?.getState();
    expect(state.valid).not.toBe(true);
  });

  test('a checkbox should not be selected if value is not on', async () => {
    const f = {
      ...field,
      enum: [false, true],
      value: true,
    };
    const { renderResponse, element } = await helper(f);
    const input = await renderResponse.container.getElementsByClassName("cmp-adaptiveform-checkbox__widget");
    expect(input).not.toBeNull();
    const state = element?.getState();
    expect(state.valid).not.toBe(true);
  });

  test('a required checkbox with no off value should be invalid when unchecked', async () => {
    const f = {
      ...field,
      enum: [true],
      required: true,
    };
    const { renderResponse, element } = await helper(f);
    const input = await renderResponse.container.getElementsByClassName("cmp-adaptiveform-checkbox__widget");
    userEvent.click(input[0]);
    userEvent.click(input[0]);
    let state = element?.getState();
    expect(state.valid).toBe(false);
  });

  test('html in the label should be handled for non rich text', async () => {
    const f = {
      ...field,
      label: {
        value: '<p>title inside p tags</p>',
        richText: true,
        visible: true
      }
    }
    let { renderResponse } = await helper(f);
    expect(renderResponse.container.innerHTML).toContain('<p>title inside p tags</p>');
  });

});