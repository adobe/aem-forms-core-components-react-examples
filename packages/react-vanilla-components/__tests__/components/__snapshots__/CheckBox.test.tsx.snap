// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Snapshot 1`] = `
<div>
  <div
    class="cmp-adaptiveform-checkbox cmp-adaptiveform-checkbox--empty "
    data-cmp-enabled="true"
    data-cmp-is="adaptiveFormCheckBox"
    data-cmp-visible="true"
  >
    <div
      class="cmp-adaptiveform-checkbox__help-container"
    />
    <div
      class="cmp-adaptiveform-checkbox__widget-container"
    >
      <input
        aria-checked="false"
        class="cmp-adaptiveform-checkbox__widget"
        id="W1yNs1AsHf3-widget"
        name="checkbox"
        type="checkbox"
        value=""
      />
      <label
        class="cmp-adaptiveform-checkbox__label"
        for="W1yNs1AsHf3-widget"
      >
        CheckBox Button
      </label>
    </div>
    <div
      aria-live="polite"
    />
  </div>
</div>
`;
